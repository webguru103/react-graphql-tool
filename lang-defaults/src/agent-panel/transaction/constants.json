[
  {
    "id": "error-renaming-transaction",
    "defaultMessage": "Sorry, error renaming transaction!"
  },
  {
    "id": "error-deleting-transaction",
    "defaultMessage": "Sorry, error deleting transaction!"
  },
  {
    "id": "add-new-name",
    "defaultMessage": "Please add a new name"
  },
  {
    "id": "rename-dialog-submit",
    "defaultMessage": "Done"
  },
  {
    "id": "rename-item-dialog-header",
    "defaultMessage": "Rename Item:"
  },
  {
    "id": "table-head-name",
    "defaultMessage": "Name"
  },
  {
    "id": "table-head-owner",
    "defaultMessage": "Owner"
  },
  {
    "id": "table-head-last-modified",
    "defaultMessage": "Last Modified"
  },
  {
    "id": "menu-rename",
    "defaultMessage": "Rename"
  },
  {
    "id": "menu-copy",
    "defaultMessage": "Copy"
  },
  {
    "id": "menu-delete",
    "defaultMessage": "Delete"
  },
  {
    "id": "empty-transaction-name",
    "defaultMessage": "Enter transaction name"
  },
  {
    "id": "no-form-selected",
    "defaultMessage": "Select at least one form"
  }
]