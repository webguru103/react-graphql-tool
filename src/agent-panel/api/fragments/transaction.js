import gql from 'graphql-tag';
import {
  transactionSessionBasicFragment,
  transactionSessionDocsFragment,
  transactionSessionSigneesFragment,
} from './transactionSession';
import type { TransactionSession } from './transactionSession';

export type TransactionBasic = {
  id: number,
  transactionName: ?string,
  createdBy: ?number,
  createdAt: ?Date,
};

export type Transaction = {
  ...$Exact<TransactionBasic>,

  transactionSessionsByTransactionId: {
    nodes: Array<?TransactionSession>
  },
}

export const transactionBasicFragment = gql`
  fragment transactionBasic on Transaction {
    id
    transactionName
    createdBy
    createdAt
  }
`;

export const transactionFragment = gql`
  fragment transaction on Transaction {
    ...transactionBasic
    # add user
    
    transactionSessionsByTransactionId {
      nodes {
        ...transactionSessionBasic
        ...transactionSessionDocs
        ...transactionSessionSignees
      }
    }
  }
  ${transactionBasicFragment}
  ${transactionSessionBasicFragment}
  ${transactionSessionSigneesFragment}
  ${transactionSessionDocsFragment}
`;
