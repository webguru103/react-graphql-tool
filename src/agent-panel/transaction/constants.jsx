import React from 'react';
import { FormattedMessage } from 'react-intl';

// TODO: create utility function to create FormattedMessages on demand, so all these are not stored in memory
// eg. to call translated text, just call function with id: trans('id', 'd'efaultMessage'), and will return:
// <FormattedMessage id="id" defaultMessage="defaultMessage" />

const DEFAULT_PAGE_SIZE = 15;

export default {
  DEFAULT_PAGE_SIZE,
  ERROR_RENAMING_TRANSACTION: <FormattedMessage id="error-renaming-transaction" defaultMessage="Sorry, error renaming transaction!" />,
  ERROR_DELETING_TRANSACTION: <FormattedMessage id="error-deleting-transaction" defaultMessage="Sorry, error deleting transaction!" />,
  ADD_NEW_NAME: <FormattedMessage id="add-new-name" defaultMessage="Please add a new name" />,
  RENAME_DIALOG_SUBMIT: <FormattedMessage id="rename-dialog-submit" defaultMessage="Done" />,
  RENAME_ITEM_DIALOG_HEADER: <FormattedMessage id="rename-item-dialog-header" defaultMessage="Rename Item:" />,
  TABLE_HEAD_NAME: <FormattedMessage id="table-head-name" defaultMessage="Name" />,
  TABLE_HEAD_OWNER: <FormattedMessage id="table-head-owner" defaultMessage="Owner" />,
  TABLE_HEAD_LAST_MODIFIED: <FormattedMessage id="table-head-last-modified" defaultMessage="Last Modified" />,
  MENU_RENAME: <FormattedMessage id="menu-rename" defaultMessage="Rename" />,
  MENU_COPY: <FormattedMessage id="menu-copy" defaultMessage="Copy" />,
  MENU_DELETE: <FormattedMessage id="menu-delete" defaultMessage="Delete" />,
  ENTER_TRANSACTION_NAME: <FormattedMessage id="empty-transaction-name" defaultMessage="Enter transaction name" />,
  NO_FORM_SELECTED: <FormattedMessage id="no-form-selected" defaultMessage="Select at least one form" />,
};
