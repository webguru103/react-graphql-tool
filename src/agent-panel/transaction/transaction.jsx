import React from 'react';
import { withRouter } from 'react-router-dom';
import type { Match } from 'react-router';
import CircularProgress from '@material-ui/core/CircularProgress';
import { withStyles } from '@material-ui/core/styles';
import { withTransactions } from './api/transaction.service';
import { compose } from '../../utility';
import Content from './contentTransaction/contentTransaction';
import styles from './transaction.styles';
import { DEBOUNCE_TIMEOUT } from '../../configurations';
import type { TransactionType } from './flowTypes';

type Props = {
  classes: Object,
  fetchMore: Function,
  transactions: Array<TransactionType>,
  history: Object,
  match: Match,
  onFilter: Function,
  filterInput: string,
}

export class TransactionC extends React.Component<Props, *> {

  static getDerivedStateFromProps(nextProps: Props, prevState: any) {
    if (nextProps.transactions !== prevState.visibleTransactions) {
      return ({ visibleTransactions: nextProps.transactions });
    }
    return null;
  }

  state = {
    loadingMore: false,
    filtering: false,
    visibleTransactions: this.props.transactions,
    filterInput: '',
    page: 0,
  };

  containerRef = null;

  componentDidMount() {
    if (this.containerRef) {
      window.addEventListener('scroll', this.fetchMoreOnScroll);
    }
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.filterInput !== this.props.filterInput) {
      this.handleTransactionsFilter(this.props.filterInput);
    }
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.fetchMoreOnScroll);
  }

  onTransactionClick = (id: string, name: string) => {
    const { history, match } = this.props;
    history.push({ pathname: `${match.path}/${id}`, state: { ...history.state, transactionName: name, transactionId: id } });
  };

  fetchMoreOnScroll = () => {
    if (
      document.documentElement && document.body
      && (document.body.offsetHeight + document.documentElement.scrollTop) >= document.documentElement.scrollHeight) {
      if (!this.state.loadingMore) {
        this.setState({ loadingMore: true });
        setTimeout(async () => {
          await this.props.fetchMore({ page: this.state.page + 1, filter: this.state.filterInput });
          this.setState(prevState => ({ loadingMore: false, page: prevState.page + 1 }));
        }, DEBOUNCE_TIMEOUT);
      }
    }
  };

  handleTransactionsFilter = (filter: string) => {
    this.setState({ filtering: true, filterInput: filter, page: 0 });
    setTimeout(async () => {
      await this.props.onFilter({ filter: this.state.filterInput });
      this.setState({ filtering: false });
    }, DEBOUNCE_TIMEOUT);
  };

  render() {
    const {
      visibleTransactions, loadingMore, filtering, filterInput,
    } = this.state;
    const { classes } = this.props;
    return (
      <div ref={r => (this.containerRef = r)} className={classes.transaction}>
        <Content
          transactions={visibleTransactions}
          onTransactionClick={this.onTransactionClick}
          filtering={filtering}
          filterInput={filterInput}
        />
        {filtering
          && (
          <div className={classes.filteringProgress}>
            <CircularProgress />
          </div>
          )
        }
        <div className={classes.loadMore}>
          {loadingMore && <CircularProgress />}
        </div>
      </div>
    );
  }

}

export default compose(
  withStyles(styles, { withTheme: true }),
  withTransactions,
  withRouter,
)(TransactionC);
