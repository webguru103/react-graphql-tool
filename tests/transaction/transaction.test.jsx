import React from 'react';
import { TransactionC } from '../../src/agent-panel/transaction/transaction';
import { shallowWrap, mountWrap } from '../test-helpers/contextWrapper';
import transactionMockData from '../test-helpers/transactionMockData';

describe('TransactionC', () => {
  let props;
  let component;
  const wrappedShallow = (lang = 'en') => {
    component = shallowWrap(<TransactionC {...props} />, lang);
    return component;
  };

  const wrappedMount = (lang = 'en') => {
    component = mountWrap(<TransactionC {...props} />, lang);
    return component;
  };

  beforeEach(() => {
    props = {
      classes: {},
      fetchMore: jest.fn(),
      transactions: transactionMockData,
      history: {
        push: jest.fn(),
      },
      match: {
        path: '/agent/transactions',
      },
      onFilter: jest.fn(),
      createDialog: jest.fn(),
    };
    if (component) component.unmount();
  });

  test('should render', () => {
    const wrapper = wrappedShallow();
    expect(wrapper).toMatchSnapshot();
  });

  test('should pass in correct props', () => {
    const wrapper = wrappedMount();
    expect(wrapper.find('TransactionC').props().transactions).toBe(transactionMockData);
  });

  test('should find row name on page for each row passed in', () => {
    const wrapper = wrappedMount();
    transactionMockData.forEach((item) => {
      expect(wrapper.contains(item.name)).toBe(true);
    });
  });

});
