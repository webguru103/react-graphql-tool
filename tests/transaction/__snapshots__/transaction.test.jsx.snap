// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionC should render 1`] = `
<div>
  <WithStyles(Apollo(CopyTransactionHOC))
    filterInput=""
    filtering={false}
    onTransactionClick={[Function]}
    transactions={
      Array [
        Object {
          "createdAt": "01/01/2018",
          "id": "id1",
          "name": "Example 1",
          "updatedAt": "01/01/2018",
        },
        Object {
          "createdAt": "01/01/2018",
          "id": "id2",
          "name": "Example Deal 2",
          "updatedAt": "01/01/2018",
        },
        Object {
          "createdAt": "01/01/2015",
          "id": "id3",
          "name": "Example 3",
          "updatedAt": "01/01/2018",
        },
        Object {
          "createdAt": "01/01/2015",
          "id": "id4",
          "name": "Example 4",
          "updatedAt": "01/01/2018",
        },
      ]
    }
  />
  <div />
</div>
`;
